### PR TITLE
Update submodule to ping-cpp

### PR DESCRIPTION
**Pull request by jaxxzer**
Before merging this:
- merge ping-protocol pr: https://github.com/bluerobotics/ping-protocol/pull/156
- update ping-protocol submodule in ping-cpp and force push (so that we use a commit from ping-protocol master)
- update the ping-cpp submodule here (so that we use a commit from ping-cpp master)

### CODE REVIEW

**Comment by patrickelectric** [comment]
Use device-manager, master branch is only for bug fixes right now related to ping1D.
It would be safer to move that in the actual wip branch.

File: qml/MainMenu.qml [hunk 1]
```diff
@@ -6,7 +6,6 @@ import QtQuick.Controls.Material 2.2
 import QtQuick.Layouts 1.3
 import Qt.labs.settings 1.0
 
-import Ping1DNamespace 1.0
```
Comment (patrickelectric) [inline qml/MainMenu.qml]: This breaks "Emit Ping" button.

File: src/sensor/parser-ping.cpp [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+        PingParser::ParseState state = parser_.parseByte(data.at(i));
+        if (state == PingParser::ParseState::NEW_MESSAGE) {
+            parsed++;
+            emit newMessage(ping_message(parser_.rxMessage));
+        } else if (state == PingParser::ParseState::ERROR) {
+            errors++;
+            emit parseError();
+        }
+    }
+}
```
Comment (patrickelectric) [inline src/sensor/parser-ping.cpp]: This function does not return any type of information related to the correct parser of a message, it should return a bool or a state condition to allow synchronous logic.

File: src/sensor/parser.h [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    Q_OBJECT
+public:
+    Parser() = default;
+    ~Parser() = default;
+
+    uint32_t parsed; // number of messages/packets successfully parsed
+    uint32_t errors; // number of parse errors
+
+    // asynchronous use, Child classes should signal when something happens ie. 'emit newMessage(Message m)'
+    virtual void parseBuffer(const QByteArray& data) { Q_UNUSED(data) }
```
Comment (patrickelectric) [inline src/sensor/parser.h]: This should be pure virtual

File: src/sensor/parser.h [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    // synchronous use, Child should return flags indicating incremental parse result/status
+    virtual uint8_t parseByte(const char byte) { Q_UNUSED(byte) return 0; }
```
Comment (patrickelectric) [inline src/sensor/parser.h]: This should be pure virtual

File: src/sensor/parser.h [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
```
Comment (patrickelectric) [inline src/sensor/parser.h]: The class should return somehow a synchronous way to get the last parsed message.

File: src/sensor/parser.h (hunk 4 above)
Comment (patrickelectric) [inline src/sensor/parser.h]: Why is this function returning an `uint8_t` ?

File: src/sensor/parser.h [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+signals:
+    void newMessage(ping_message msg);
```
Comment (patrickelectric) [inline src/sensor/parser.h]: `const ping_message&`

File: pingviewer.pro [hunk 7]
```diff
@@ -84,4 +84,4 @@ exists ($$_PRO_FILE_PWD_/.git) {
 message(" GIT_VERSION:      " $$GIT_VERSION)
 message(" GIT_VERSION_DATE: " $$GIT_VERSION_DATE)
 message(" GIT_TAG:          " $$GIT_TAG)
-message(" GIT_URL:          " $$GIT_URL)
\ No newline at end of file
+message(" GIT_URL:          " $$GIT_URL)
```
Comment (patrickelectric) [inline pingviewer.pro]: Remove style changes.

File: src/link/pingsimulationlink.cpp [hunk 8]
```diff
@@ -1,7 +1,9 @@
 #include <QtMath>
 
 #include "pingsimulationlink.h"
-#include "pingmessage/ping_ping1d.h"
+#include <ping-message.h>
```
Comment (patrickelectric) [inline src/link/pingsimulationlink.cpp]: Why is this include here ?

File: src/link/pingsimulationlink.cpp [hunk 9]
```diff
@@ -1,7 +1,9 @@
...
+#include <ping-message-common.h>
```
Comment (patrickelectric) [inline src/link/pingsimulationlink.cpp]: The class is not using common messages

File: src/link/simulationlink.h [hunk 10]
```diff
@@ -23,7 +23,7 @@ class SimulationLink : public AbstractLink
      * @return true
      * @return false
      */
-    bool isWritable() final { return false; };
+    bool isWritable() final { return false; }
```
Comment (patrickelectric) [inline src/link/simulationlink.h]: Avoid style changes

File: src/main.cpp [hunk 11]
```diff
@@ -58,11 +55,6 @@ int main(int argc, char *argv[])
     qmlRegisterType<Ping>("Ping", 1, 0, "Ping");
     qmlRegisterType<Waterfall>("Waterfall", 1, 0, "Waterfall");
 
-    // Uncreatable register
-    qmlRegisterUncreatableMetaObject(
-        PingPing1DNamespace::staticMetaObject, "Ping1DNamespace", 1, 0, "Ping1DNamespace", "This is a enum."
```
Comment (patrickelectric) [inline src/main.cpp]: This is used in some qml files.

File: src/sensor/ping.h [hunk 12]
```diff
@@ -253,9 +253,9 @@ class Ping : public Sensor
      *
      * @param id
      */
-    void do_continuous_start(PingPing1DNamespace::ping_ping1D_id id)
+    void do_continuous_start(uint16_t id)
```
Comment (patrickelectric) [inline src/sensor/ping.h]: This is a high level class, it's not save to use uint16_t as parameter, use enum.
The enum is explicit and allows the compiler to detect problems.

File: src/sensor/ping.h [hunk 13]
```diff
@@ -648,5 +648,5 @@ class Ping : public Sensor
         int waiting = 0;
     };
     friend QDebug operator<<(QDebug d, const Ping::messageStatus& other);
-    QHash<PingPing1DNamespace::ping_ping1D_id, messageStatus> requestedIds;
+    QHash<uint16_t, messageStatus> requestedIds;
```
Comment (patrickelectric) [inline src/sensor/ping.h]: This is a high level class, it's not save to use uint16_t as key, use enum.
The enum is explicit and allows the compiler to detect problems.

File: src/test.cpp [hunk 14]
```diff
@@ -171,4 +170,4 @@ void Test::waterfallGradient()
 
 }
 
-QTEST_MAIN(Test)
```
Comment (patrickelectric) [inline src/test.cpp]: Avoid style changes.

File: src/sensor/parser.h [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+    void parseError();
+};
```
Comment (patrickelectric) [inline src/sensor/parser.h]: This should contain a `clearBuffer()` function to remove any trace of old information from older tries for mult-sensor logic.

File: qml/MainMenu.qml (hunk 1 above)
Comment (jaxxzer) [reply qml/MainMenu.qml]: @patrick I think instead of exporting pingmessage enums to qml, it's better to have a QINVOKABLE emitPing()

File: src/sensor/parser-ping.cpp (hunk 2 above)
Comment (jaxxzer) [reply src/sensor/parser-ping.cpp]: This function is explicitly for asynchronous use only: https://github.com/bluerobotics/ping-viewer/blob/050fe06c7bba575b5a824c9406eba6c1aca2a476/src/sensor/parser.h#L17

Synchronous operation is only allowed with the parseByte() function  (as in arduino implementation).

Comment (jaxxzer) [reply src/sensor/parser-ping.cpp]: I've updated this class to allow synchronous operation according to your recommendation, but this function remains void return.

File: src/sensor/parser.h (hunk 15 above)
Comment (jaxxzer) [reply src/sensor/parser.h]: done

File: src/test.cpp (hunk 14 above)
Comment (jaxxzer) [reply src/test.cpp]: Our guideline is a newline at the end of the file.

File: qml/MainMenu.qml (hunk 1 above)
Comment (jaxxzer) [reply qml/MainMenu.qml]: I added emitPing(), it's less code for the whole project.

File: src/sensor/parser.h (hunk 3 above)
Comment (jaxxzer) [reply src/sensor/parser.h]: I have a problem when I attempt to make it pure virtual:
Screenshot_20190621_094559

File: src/link/simulationlink.h (hunk 10 above)
Comment (jaxxzer) [reply src/link/simulationlink.h]: This is your preference (no semi colon). You want a new commit?

File: src/sensor/parser.h (hunk 4 above)
Comment (jaxxzer) [reply src/sensor/parser.h]: fixed

File: src/sensor/parser.h (hunk 5 above)
Comment (jaxxzer) [reply src/sensor/parser.h]: done

File: src/sensor/parser.h (hunk 6 above)
Comment (jaxxzer) [reply src/sensor/parser.h]: done

File: src/link/simulationlink.h (hunk 10 above)
Comment (jaxxzer) [reply src/link/simulationlink.h]: I'll put the style changes in different commits.

File: src/main.cpp (hunk 11 above)
Comment (jaxxzer) [reply src/main.cpp]: Not anymore!

File: src/sensor/parser.h (hunk 15 above)
Comment (jaxxzer) [reply src/sensor/parser.h]: done

File: src/link/pingsimulationlink.cpp (hunk 8 above)
Comment (jaxxzer) [reply src/link/pingsimulationlink.cpp]: removed

File: src/link/pingsimulationlink.cpp (hunk 9 above)
Comment (jaxxzer) [reply src/link/pingsimulationlink.cpp]: removed

File: src/sensor/ping.h (hunk 13 above)
Comment (jaxxzer) [reply src/sensor/ping.h]: a uint16_t is safe by definition - a message id is a uint16_t. To change the signature here will also require a single enum scope for all message ids (common, ping1d, ping360), which we do not have (we could make it if we require reserved ids for each sensor).

You can still gain the compiler/enum advantages: I recommend the caller make use of the the individual message id enum/namespaces.

File: src/sensor/ping.h (hunk 12 above)
Comment (jaxxzer) [reply src/sensor/ping.h]: done

File: src/link/simulationlink.h (hunk 10 above)
Comment (jaxxzer) [reply src/link/simulationlink.h]: I'm dropping them for expediency.

File: src/test.cpp (hunk 14 above)
Comment (jaxxzer) [reply src/test.cpp]: removed the change

File: src/sensor/parser.h (hunk 3 above)
Comment (jaxxzer) [reply src/sensor/parser.h]: I assume it's because we have Parser member in abstract link, it the functions need to be defined.

**Comment by jaxxzer** [comment]
Update, all review has been addressed, and everything is working with the application locally. I am working now on adding class documentation for the parsers and fixing the ci fail.

**Comment by jaxxzer** [comment]
The application works with ping1d simulation, but not with a real ping1d hardware. It looks like some problem with the request timer callback and/or the initializing sensor settings from previous session.

File: src/sensor/parser.h (hunk 5 above)
Comment (patrickelectric) [reply src/sensor/parser.h]: Why `parseBuffer` does not return state ? This will obligate us to create a loop and do what `parBuffer` is already doing.

File: src/sensor/parser.h [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+    virtual ParserState parseByte(const char byte) { Q_UNUSED(byte) return ERROR; }
+
+    ping_message rxMessage() { return rxMessage_; }
+
+signals:
+    void newMessage(const ping_message& msg);
+    void parseError();
+
+protected:
+    ping_message rxMessage_;
```
Comment (patrickelectric) [inline src/sensor/parser.h]: Please follow our code style:
https://github.com/bluerobotics/software-guidelines/blob/master/guidelines/style.md#code

File: src/sensor/parser.h [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+
+    //TODO move to pure virtual
+    // clear parse state
+    virtual void clearBuffer() { }
+    // asynchronous use, Child classes should signal when something happens ie. 'emit newMessage(Message m)'
+    virtual void parseBuffer(const QByteArray& data) { Q_UNUSED(data) }
+    // synchronous use, Child should return flags indicating incremental parse result/status
+    virtual ParserState parseByte(const char byte) { Q_UNUSED(byte) return ERROR; }
+
+    ping_message rxMessage() { return rxMessage_; }
```
Comment (patrickelectric) [inline src/sensor/parser.h]: const function.

File: src/sensor/ping.h [hunk 18]
```diff
@@ -185,23 +185,28 @@ class Ping : public PingSensor
...
         m.updateChecksum();
         writeMessage(m);
-        request(Ping1DNamespace::ModeAuto);
+        request(Ping1dId::MODE_AUTO);
     }
     Q_PROPERTY(bool mode_auto READ mode_auto WRITE set_mode_auto NOTIFY modeAutoUpdate)
 
+    // sensor only supports continuous profile output
+    enum ContinuousId : uint16_t {
+        PROFILE = Ping1dId::PROFILE
```
Comment (patrickelectric) [inline src/sensor/ping.h]: This doesn't look as a good approach, if `do_continuous_start` only supports one argument, it should not accept any argument at all.

File: src/sensor/ping.h (hunk 13 above)
Comment (patrickelectric) [reply src/sensor/ping.h]: From the creator of C++ language:

> enum class -- scoped and strongly typed enums
The enum classes ("new enums", "strong enums") address three problems with traditional C++ enumerations:
conventional enums implicitly convert to int, causing errors when someone does not want an enumeration to act as an integer.
conventional enums export their enumerators to the surrounding scope, causing name clashes.
the underlying type of an enum cannot be specified, causing confusion, compatibility problems, and makes forward declaration impossible.

http://www.stroustrup.com/C++11FAQ.html#enum

There is also the cpp guidelines:
https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#enum-enumerations

File: src/sensor/parser-ping.cpp [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+void PingParserExt::clearBuffer()
+{
+    parser_.reset();
+}
+
+void PingParserExt::parseBuffer(const QByteArray& data)
+{
+    for(int i = 0; i < data.length(); i++) {
+        PingParser::ParseState state = parser_.parseByte(data.at(i));
+        if (state == PingParser::ParseState::NEW_MESSAGE) {
```
Comment (patrickelectric) [inline src/sensor/parser-ping.cpp]: Why ParseState is necessary ?

File: src/sensor/parser-ping.h [hunk 20]
```diff
@@ -0,0 +1,16 @@
+#pragma once
+
+#include "parser.h"
+#include "ping-parser.h"
+
+class PingParserExt : public Parser
+{
+public:
+    PingParserExt() : parser_(2048) {}
```
[the remark this answers is not in the document]
Comment (jaxxzer) [reply src/sensor/parser-ping.h]: undefined behavior

File: src/sensor/parser-ping.h [hunk 21]
```diff
@@ -0,0 +1,16 @@
...
+    void clearBuffer() override final;
+    void parseBuffer(const QByteArray& data) override final;
+    ParserState parseByte(const char byte) override final;
+
+private:
+    PingParser parser_;
```
[the remark this answers is not in the document]
Comment (jaxxzer) [reply src/sensor/parser-ping.h]: It will take some time, but I can change it

File: src/sensor/parser.h (hunk 3 above)
Comment (jaxxzer) [reply src/sensor/parser.h]: ok

File: src/sensor/parser.h (hunk 16 above)
Comment (jaxxzer) [reply src/sensor/parser.h]: I was reading some things that prefixing with underscores is not recommended for c/c++, it has some reserved or special meaning.

It will take some time, but I can change it.

File: src/sensor/ping.h (hunk 18 above)
Comment (jaxxzer) [reply src/sensor/ping.h]: The idea is to use it to request different messages, so it is the correct design for `do_continuous_start` to have a message id parameter. The sensor just doesn't support continuous output for `distance` messages yet. When it does, we will add it to the list here.

File: src/sensor/ping.h (hunk 13 above)
Comment (jaxxzer) [reply src/sensor/ping.h]: Thank you, I've read it and I understand the enum, the enum class, type safety and the advantages of using them. I still think it is the wrong idea to have an enum type for the parameter. (But I think it is the right idea for the caller to always use enum types for the arguments! ie `request(Ping1dId::DISTANCE)`. You can still write your code using enum types in every call.

Because you can, does not mean you should. Because it compiles, does not make it right. It is similar to our past conversation about if we should declare something const here:  https://github.com/bluerobotics/ping-protocol/pull/77

File: pingviewer.pro [hunk 22]
```diff
@@ -22,8 +22,8 @@ RESOURCES += \
     QMAKE_LFLAGS += -fopenmp
 }
 
-include(lib/ping-protocol-cpp/ping.pri)
 include(lib/maddy/maddy.pri)
+include(lib/lib.pri)
```
[the remark this answers is not in the document]
Comment (jaxxzer) [reply pingviewer.pro]: That seems unnecessary. What is the problem with this? Do you have some reference for the standard?

File: src/sensor/parser-ping.cpp (hunk 19 above)
Comment (jaxxzer) [reply src/sensor/parser-ping.cpp]: It is to expose the current state of the parser. There are **three** possible states of interest: new message, parsing, and error.

File: src/sensor/parser-ping.cpp (hunk 2 above)
Comment (jaxxzer) [reply src/sensor/parser-ping.cpp]: Because `parseBuffer` might parse more than one message in a single call. It was created to add async signalling to the class.

Comment (jaxxzer) [reply src/sensor/parser-ping.cpp]: We could make it bool return, but I don't see any use case to apply it to.

**Comment by jaxxzer** [comment]
> The application works with ping1d simulation, but not with a real ping1d hardware.

update: my bug working with the real hardware is fixed

File: pingviewer.pro (hunk 22 above)
[the remark this answers is not in the document]
Comment (jaxxzer) [reply pingviewer.pro]: done, but I don't like the extra directory level to hold just a .pri.

File: src/sensor/parser-ping.h (hunk 20 above)
[the remark this answers is not in the document]
Comment (jaxxzer) [reply src/sensor/parser-ping.h]: I added some documentation for the constructor

**Comment by jaxxzer** [comment]
There are two remaining things here it will help me if you can take a look @patrickelectric, I can't see how my patch has affected these:
- the lost messages is initialized with a value of ~500k
- the control panel is not working from some undefined reference to ping in qml

File: lib/lib.pri [hunk 23]
```diff
@@ -0,0 +1,5 @@
+MESSAGEPATH = $$PWD/ping-cpp/src/message
+
+HEADERS += $$MESSAGEPATH/*.h
+
+INCLUDEPATH += $$MESSAGEPATH
```
Comment (patrickelectric) [inline lib/lib.pri]: This file is not necessary anymore.

File: src/sensor/parser-ping.cpp (hunk 19 above)
Comment (patrickelectric) [reply src/sensor/parser-ping.cpp]: I'm saying that in some points you use `PingParser::NEW_MESSAGE`, `PingParser::ParseState::NEW_MESSAGE` and `Parser::ParserState::NEW_MESSAGE`.
There is two different classes and the enum name is not being used in some places resulting in 3 different enumeration uses. 
Is it possible to use the base class enum `Parser` with/without `ParserState` to help with readability and maintenance and avoiding the usage of a secondary kind ?

File: src/sensor/parser-ping.cpp [hunk 24]
```diff
@@ -0,0 +1,34 @@
...
+}
+
+Parser::ParserState PingParserExt::parseByte(const char byte)
+{
+    uint8_t state = _parser.parseByte(byte);
+    if (state == PingParser::ParseState::NEW_MESSAGE) {
+        _rxMessage = _parser.rxMessage;
+        return Parser::ParserState::NEW_MESSAGE;
+    } else if (state == PingParser::ERROR) {
+        return Parser::ParserState::ERROR;
```
Comment (patrickelectric) [inline src/sensor/parser-ping.cpp]: Why we are not using the base class enum here ? What is the problem to use PingParser ?

File: src/sensor/parser-ping.h [hunk 25]
```diff
@@ -0,0 +1,38 @@
...
+ * @brief The PingParserExt class wraps the PingParser class from the ping-protocol submodule
+ * and Extends it with signalling in order to subclass our Parser class
+ */
+class PingParserExt : public Parser
+{
+public:
+    /**
+     * @brief Any messages parsed must be shorter than the buffer length
+     */
+    PingParserExt() : _parser(2048) {}
```
Comment (patrickelectric) [inline src/sensor/parser-ping.h]: Take in mind that Ping360 `device_data` can use up to 2026 points (if the max profile length is 2000 points).
This is not a microcontroller, we can put 10240 here just to be safe, or if you want to be safer and avoid multiple page memories, you can use 4096.

File: src/sensor/ping.h [hunk 26]
```diff
@@ -321,6 +329,11 @@ class Ping : public PingSensor
     Q_INVOKABLE void firmwareUpdate(QString fileUrl, bool sendPingGotoBootloader = true, int baud = 57600,
                                     bool verify = true);
 
+    /**
+     * @brief request a profile message from the device
+     */
+    Q_INVOKABLE void emitPing() { request(Ping1dId::PROFILE); }
```
Comment (patrickelectric) [inline src/sensor/ping.h]: const function.

File: src/sensor/pingsensor.cpp [hunk 27]
```diff
@@ -22,8 +24,9 @@ void PingSensor::request(int id)
     }
     qCDebug(PING_PROTOCOL_PINGSENSOR) << "Requesting:" << id;
 
-    ping_message_empty m;
-    m.set_id(id);
+    ping_message m(10);
```
Comment (patrickelectric) [inline src/sensor/pingsensor.cpp]: Why 10 is being used here ?
Should 10 be the default constructor ?

File: src/sensor/pingsensor.cpp [hunk 28]
```diff
@@ -22,8 +24,9 @@ void PingSensor::request(int id)
...
+    m.set_payload_length(0);
```
Comment (patrickelectric) [inline src/sensor/pingsensor.cpp]: Wait, why is this necessary ? Why is not zero by default ?

File: qml/Ping1DControlPanel.qml [hunk 29]
```diff
@@ -5,8 +5,6 @@ import QtQuick.Controls 2.2
 import QtQuick.Controls.Material 2.2
 import QtQuick.Layouts 1.3
 
-import DeviceManager 1.0
```
Comment (patrickelectric) [inline qml/Ping1DControlPanel.qml]: This is necessary to get the connected device, otherwise the entire widget will not work.

File: src/sensor/parser.h [hunk 30]
```diff
@@ -0,0 +1,60 @@
...
+     * @return the last message that was parsed
+     */
+    ping_message rxMessage() const { return _rxMessage; }
+
+signals:
+    void newMessage(const ping_message& msg);
+    void parseError();
+
+protected:
+    ping_message _rxMessage;
```
Comment (patrickelectric) [inline src/sensor/parser.h]: Parser should be a generic class, no ? So why there is ping_message here ?
Should this be in PingParserExt ?

File: qml/Ping1DControlPanel.qml (hunk 29 above)
Comment (jaxxzer) [reply qml/Ping1DControlPanel.qml]: uuuuuff how did I miss this

File: src/sensor/pingsensor.cpp (hunk 27 above)
Comment (jaxxzer) [reply src/sensor/pingsensor.cpp]: This is to construct an arbitrary pingmessage by hand.
Here we are making our 'empty' pingmessage with a message length of 10.

> Should 10 be the default constructor ?

Maybe. If we add 10 as default argument it will conflict with another constructor, i'm not sure if that one is used. We can/should definitely try to make the pingmessage class more robust to memory access violations.

File: src/sensor/parser-ping.cpp (hunk 24 above)
Comment (jaxxzer) [reply src/sensor/parser-ping.cpp]: > Why we are not using the base class enum here ?

In our project, `Parser::ParseByte` returns the enum type `Parser::ParseState`: https://github.com/bluerobotics/ping-viewer/blob/protocol/src/sensor/parser.h#L19

In ping-cpp, `PingParser::parseByte` returns the enum type `PingParser::ParseState`: https://github.com/bluerobotics/ping-cpp/blob/deployment/src/message/ping-parser.h#L17

The base class Parser has no knowledge of the `PingParser` enum from ping-cpp. The PingParserExt method here does the conversion between the states.

Comment (jaxxzer) [reply src/sensor/parser-ping.cpp]: `Parser::ParseState::NEW_MESSAGE == PingParser::ParseState::NEW_MESSAGE` is not necessarily true, that's why the code is written like this.

File: src/sensor/pingsensor.cpp (hunk 28 above)
Comment (jaxxzer) [reply src/sensor/pingsensor.cpp]: msgData is allocated on the heap and not zero initialized, nor is it initialized by this message constructor, because the constructor doesn't know anything about the message payload or it's length.

File: src/sensor/parser.h (hunk 30 above)
Comment (jaxxzer) [reply src/sensor/parser.h]: You are right... We could change the signal to `newMessage()`, then copy the message synchronously.

File: src/sensor/ping.h (hunk 26 above)
Comment (jaxxzer) [reply src/sensor/ping.h]: done, and now Sensor::request is const too as a requirement.

File: src/sensor/parser-ping.h (hunk 25 above)
Comment (jaxxzer) [reply src/sensor/parser-ping.h]: done

File: lib/lib.pri (hunk 23 above)
Comment (jaxxzer) [reply lib/lib.pri]: done

File: src/sensor/parser.h (hunk 30 above)
Comment (jaxxzer) [reply src/sensor/parser.h]: This is a mistake from the past. I think we should leave it alone, and reevaluate it when we have the need for another parser/protocol.

**Comment by jaxxzer** [comment]
@patrickelectric I think we are close.. let me know anything else to change.

File: src/sensor/pingsensor.cpp (hunk 28 above)
Comment (jaxxzer) [reply src/sensor/pingsensor.cpp]: Our `ping_message_empty` constructor did the initialization for us manually in the past.

**Comment by jaxxzer** [comment]
The only thing I don't know why it's changed, is the lost messages is initialized at this number every time:
Screenshot_20190626_235835

**Comment by patrickelectric** [comment]
> The only thing I don't know why it's changed, is the lost messages is initialized at this number every time:
> Screenshot_20190626_235835

Fixed in https://github.com/bluerobotics/ping-viewer/pull/527

**Comment by patrickelectric** [comment]
Can we merge this ?
